File: app/csharp/tests/Ch11/TravelGuideTests.cs
using Ch11;
using LanguageExt;
using static LanguageExt.Prelude;
using Xunit;
using static Ch11.TravelGuideApp;

namespace GrokkingFP.CSharp.Tests.Ch11;

/// <summary>
/// 第11章: TravelGuide アプリケーションのテスト
/// </summary>
public class LocationIdTests
{
    [Fact]
    public void LocationId_HoldsValue()
    {
        var id = LocationId.Create("Q123");
        Assert.Equal("Q123", id.Value);
    }

    [Fact]
    public void LocationId_ToStringReturnsValue()
    {
        var id = LocationId.Create("Q123");
        Assert.Equal("Q123", id.ToString());
    }
}

public class LocationTests
{
    [Fact]
    public void Location_IsCreatedCorrectly()
    {
        var location = new Location(
            LocationId.Create("Q123"),
            "Tokyo",
            14000000
        );
        Assert.Equal("Tokyo", location.Name);
        Assert.Equal(14000000, location.Population);
    }
}

public class AttractionTests
{
    [Fact]
    public void Attraction_IsCreatedCorrectly()
    {
        var location = new Location(
            LocationId.Create("Q123"),
            "Tokyo",
            14000000
        );
        var attraction = new Attraction(
            "Tokyo Tower",
            Some("Famous landmark"),
            location
        );
        Assert.Equal("Tokyo Tower", attraction.Name);
        Assert.True(attraction.Description.IsSome);
        Assert.Equal("Famous landmark", attraction.Description.IfNone(""));
    }

    [Fact]
    public void Attraction_CanHaveNoDescription()
    {
        var location = new Location(
            LocationId.Create("Q123"),
            "Tokyo",
            14000000
        );
        var attraction = new Attraction("Tokyo Tower", None, location);
        Assert.True(attraction.Description.IsNone);
    }
}

public class TravelGuideModelTests
{
    [Fact]
    public void TravelGuide_IsCreatedCorrectly()
    {
        var location = new Location(
            LocationId.Create("Q123"),
            "Tokyo",
            14000000
        );
        var attraction = new Attraction("Tokyo Tower", None, location);
        var guide = new TravelGuide(attraction, Seq("Artist1", "Movie1"));

        Assert.Equal(2, guide.Subjects.Count);
        Assert.Contains("Artist1", guide.Subjects);
        Assert.Contains("Movie1", guide.Subjects);
    }
}

public class TestDataAccessTests
{
    [Fact]
    public async Task TestDataAccess_ReturnsCorrectData()
    {
        var testLocation = new Location(
            LocationId.Create("Q123"),
            "Test City",
            100000
        );
        var testAttractions = Seq(
            new Attraction("Test Tower", Some("A test"), testLocation)
        );
        var testArtists = Seq(
            new MusicArtist("Test Artist", Some("Pop"))
        );
        var testMovies = Seq(
            new Movie("Test Movie", Some(1000000))
        );

        var dataAccess = TestDataAccess(testAttractions, testArtists, testMovies);

        var attractions = await dataAccess.FindAttractions("Test", AttractionOrdering.ByName, 10);
        Assert.Single(attractions);
        Assert.Equal("Test Tower", attractions.First().Name);
    }

    [Fact]
    public async Task TestDataAccess_Filters()
    {
        var testLocation = new Location(
            LocationId.Create("Q123"),
            "Test City",
            100000
        );
        var testAttractions = Seq(
            new Attraction("Test Tower", None, testLocation),
            new Attraction("Other Place", None, testLocation)
        );

        var dataAccess = TestDataAccess(testAttractions, Seq<MusicArtist>(), Seq<Movie>());

        var attractions = await dataAccess.FindAttractions("Tower", AttractionOrdering.ByName, 10);
        Assert.Single(attractions);
        Assert.Equal("Test Tower", attractions.First().Name);
    }

    [Fact]
    public async Task TestDataAccess_RespectsLimit()
    {
        var testLocation = new Location(
            LocationId.Create("Q123"),
            "Test City",
            100000
        );
        var testAttractions = Seq(
            new Attraction("Tower 1", None, testLocation),
            new Attraction("Tower 2", None, testLocation),
            new Attraction("Tower 3", None, testLocation)
        );

        var dataAccess = TestDataAccess(testAttractions, Seq<MusicArtist>(), Seq<Movie>());

        var attractions = await dataAccess.FindAttractions("Tower", AttractionOrdering.ByName, 2);
        Assert.Equal(2, attractions.Count);
    }
}

public class GetTravelGuideTests
{
    [Fact]
    public async Task GetTravelGuide_ReturnsGuideForValidAttraction()
    {
        var testLocation = new Location(
            LocationId.Create("Q123"),
            "Test City",
            100000
        );
        var testAttractions = Seq(
            new Attraction("Test Tower", Some("A tower"), testLocation)
        );
        var testArtists = Seq(
            new MusicArtist("Artist1", None),
            new MusicArtist("Artist2", Some("Rock"))
        );
        var testMovies = Seq(
            new Movie("Movie1", None)
        );

        var dataAccess = TestDataAccess(testAttractions, testArtists, testMovies);
        var result = await GetTravelGuide(dataAccess, "Test");

        Assert.True(result.IsSome);
        var guide = result.IfNone(() => throw new Exception("Expected Some"));
        Assert.Equal("Test Tower", guide.Attraction.Name);
        Assert.Equal(3, guide.Subjects.Count); // 2 artists + 1 movie
    }

    [Fact]
    public async Task GetTravelGuide_ReturnsNoneWhenNotFound()
    {
        var dataAccess = TestDataAccess(Seq<Attraction>(), Seq<MusicArtist>(), Seq<Movie>());
        var result = await GetTravelGuide(dataAccess, "NonExistent");

        Assert.True(result.IsNone);
    }

    [Fact]
    public async Task GetTravelGuides_ReturnsMultipleGuides()
    {
        var testLocation1 = new Location(
            LocationId.Create("Q123"),
            "City1",
            200000
        );
        var testLocation2 = new Location(
            LocationId.Create("Q456"),
            "City2",
            100000
        );
        var testAttractions = Seq(
            new Attraction("Tower1", None, testLocation1),
            new Attraction("Tower2", None, testLocation2)
        );
        var testArtists = Seq(new MusicArtist("Artist1", None));
        var testMovies = Seq(new Movie("Movie1", None));

        var dataAccess = TestDataAccess(testAttractions, testArtists, testMovies);
        var result = await GetTravelGuides(dataAccess, "Tower", 5);

        Assert.Equal(2, result.Count);
    }
}

public class CachedDataAccessTests
{
    [Fact]
    public async Task CachedDataAccess_CachesData()
    {
        var callCount = 0;
        var testLocation = new Location(
            LocationId.Create("Q123"),
            "Test City",
            100000
        );
        var testAttractions = Seq(
            new Attraction("Test Tower", None, testLocation)
        );

        var countingDataAccess = new CountingDataAccess(
            testAttractions,
            Seq<MusicArtist>(),
            Seq<Movie>(),
            () => callCount++
        );

        var cached = CachedDataAccess(countingDataAccess);

        // First call
        _ = await cached.FindAttractions("Test", AttractionOrdering.ByName, 10);
        Assert.Equal(1, callCount);

        // Second call (should be cached)
        _ = await cached.FindAttractions("Test", AttractionOrdering.ByName, 10);
        Assert.Equal(1, callCount); // Call count should not increase

        // Different parameters
        _ = await cached.FindAttractions("Other", AttractionOrdering.ByName, 10);
        Assert.Equal(2, callCount);
    }

    private class CountingDataAccess : IDataAccess
    {
        private readonly Seq<Attraction> _attractions;
        private readonly Seq<MusicArtist> _artists;
        private readonly Seq<Movie> _movies;
        private readonly Action _onCall;

        public CountingDataAccess(
            Seq<Attraction> attractions,
            Seq<MusicArtist> artists,
            Seq<Movie> movies,
            Action onCall)
        {
            _attractions = attractions;
            _artists = artists;
            _movies = movies;
            _onCall = onCall;
        }

        public Task<Seq<Attraction>> FindAttractions(string name, AttractionOrdering ordering, int limit)
        {
            _onCall();
            var filtered = _attractions.Filter(a => a.Name.Contains(name)).Take(limit);
            return Task.FromResult(filtered);
        }

        public Task<Seq<MusicArtist>> FindArtistsFromLocation(LocationId locationId, int limit)
        {
            return Task.FromResult(_artists.Take(limit));
        }

        public Task<Seq<Movie>> FindMoviesAboutLocation(LocationId locationId, int limit)
        {
            return Task.FromResult(_movies.Take(limit));
        }
    }
}

public class ResourceTests
{
    [Fact]
    public async Task Resource_Make_CreatesResource()
    {
        var acquired = false;
        var released = false;

        var resource = Resource.Make(
            async () =>
            {
                acquired = true;
                return "resource";
            },
            async r => { released = true; }
        );

        var result = await resource.UseAsync(async r => r + "-used");

        Assert.Equal("resource-used", result);
        Assert.True(acquired);
        Assert.True(released);
    }

    [Fact]
    public async Task Resource_ReleasesEvenOnException()
    {
        var released = false;

        var resource = Resource.Make(
            async () => "resource",
            async r => { released = true; }
        );

        try
        {
            await resource.UseAsync<string>(async r => throw new Exception("Test error"));
        }
        catch
        {
            // Expected
        }

        Assert.True(released);
    }
}

public class SampleDataTests
{
    [Fact]
    public void SampleLocation_HasCorrectValues()
    {
        Assert.Equal("Tokyo", SampleLocation.Name);
        Assert.Equal(14000000, SampleLocation.Population);
    }

    [Fact]
    public void SampleAttraction_HasCorrectValues()
    {
        Assert.Equal("Tokyo Tower", SampleAttraction.Name);
        Assert.True(SampleAttraction.Description.IsSome);
    }

    [Fact]
    public void SampleArtist_HasCorrectValues()
    {
        Assert.Equal("Hikaru Utada", SampleArtist.Name);
    }

    [Fact]
    public void SampleMovie_HasCorrectValues()
    {
        Assert.Equal("Lost in Translation", SampleMovie.Name);
    }
}

File: app/fsharp/tests/Ch11/TravelGuideTests.fs
module Ch11.TravelGuideTests

open Xunit
open Ch11.TravelGuide

// ============================================
// ドメインモデルのテスト
// ============================================

[<Fact>]
let ``LocationId は値を保持する`` () =
    let id = LocationId "Q123"
    Assert.Equal("Q123", LocationId.value id)

[<Fact>]
let ``Location は正しく作成される`` () =
    let location = {
        Id = LocationId "Q123"
        Name = "Tokyo"
        Population = 14000000
    }
    Assert.Equal("Tokyo", location.Name)
    Assert.Equal(14000000, location.Population)

[<Fact>]
let ``Attraction は正しく作成される`` () =
    let location = {
        Id = LocationId "Q123"
        Name = "Tokyo"
        Population = 14000000
    }
    let attraction = {
        Name = "Tokyo Tower"
        Description = Some "Famous landmark"
        Location = location
    }
    Assert.Equal("Tokyo Tower", attraction.Name)
    Assert.Equal(Some "Famous landmark", attraction.Description)

[<Fact>]
let ``TravelGuide は正しく作成される`` () =
    let location = {
        Id = LocationId "Q123"
        Name = "Tokyo"
        Population = 14000000
    }
    let attraction = {
        Name = "Tokyo Tower"
        Description = None
        Location = location
    }
    let guide = {
        Attraction = attraction
        Subjects = ["Artist1"; "Movie1"]
    }
    Assert.Equal(2, List.length guide.Subjects)

// ============================================
// データアクセスのテスト
// ============================================

[<Fact>]
let ``testDataAccess は正しいデータを返す`` () =
    let testLocation = {
        Id = LocationId "Q123"
        Name = "Test City"
        Population = 100000
    }
    let testAttractions = [
        { Name = "Test Tower"; Description = Some "A test"; Location = testLocation }
    ]
    let testArtists = [
        { Name = "Test Artist"; Genre = Some "Pop" }
    ]
    let testMovies = [
        { Name = "Test Movie"; BoxOffice = Some 1000000 }
    ]

    let dataAccess = testDataAccess testAttractions testArtists testMovies

    let attractions =
        dataAccess.FindAttractions("Test", ByName, 10)
        |> Async.RunSynchronously
    Assert.Equal(1, List.length attractions)
    Assert.Equal("Test Tower", attractions.[0].Name)

[<Fact>]
let ``testDataAccess はフィルタリングする`` () =
    let testLocation = {
        Id = LocationId "Q123"
        Name = "Test City"
        Population = 100000
    }
    let testAttractions = [
        { Name = "Test Tower"; Description = None; Location = testLocation }
        { Name = "Other Place"; Description = None; Location = testLocation }
    ]

    let dataAccess = testDataAccess testAttractions [] []

    let attractions =
        dataAccess.FindAttractions("Tower", ByName, 10)
        |> Async.RunSynchronously
    Assert.Equal(1, List.length attractions)

[<Fact>]
let ``testDataAccess は limit を尊重する`` () =
    let testLocation = {
        Id = LocationId "Q123"
        Name = "Test City"
        Population = 100000
    }
    let testAttractions = [
        { Name = "Tower 1"; Description = None; Location = testLocation }
        { Name = "Tower 2"; Description = None; Location = testLocation }
        { Name = "Tower 3"; Description = None; Location = testLocation }
    ]

    let dataAccess = testDataAccess testAttractions [] []

    let attractions =
        dataAccess.FindAttractions("Tower", ByName, 2)
        |> Async.RunSynchronously
    Assert.Equal(2, List.length attractions)

// ============================================
// travelGuide 関数のテスト
// ============================================

[<Fact>]
let ``travelGuide は有効なアトラクションに対してガイドを返す`` () =
    let testLocation = {
        Id = LocationId "Q123"
        Name = "Test City"
        Population = 100000
    }
    let testAttractions = [
        { Name = "Test Tower"; Description = Some "A tower"; Location = testLocation }
    ]
    let testArtists = [
        { Name = "Artist1"; Genre = None }
        { Name = "Artist2"; Genre = Some "Rock" }
    ]
    let testMovies = [
        { Name = "Movie1"; BoxOffice = None }
    ]

    let dataAccess = testDataAccess testAttractions testArtists testMovies
    let result = travelGuide dataAccess "Test" |> Async.RunSynchronously

    Assert.True(result.IsSome)
    let guide = result.Value
    Assert.Equal("Test Tower", guide.Attraction.Name)
    Assert.Equal(3, List.length guide.Subjects) // 2 artists + 1 movie

[<Fact>]
let ``travelGuide は見つからない場合 None を返す`` () =
    let dataAccess = testDataAccess [] [] []
    let result = travelGuide dataAccess "NonExistent" |> Async.RunSynchronously
    Assert.True(result.IsNone)

[<Fact>]
let ``travelGuides は複数のガイドを返す`` () =
    let testLocation1 = {
        Id = LocationId "Q123"
        Name = "City1"
        Population = 200000
    }
    let testLocation2 = {
        Id = LocationId "Q456"
        Name = "City2"
        Population = 100000
    }
    let testAttractions = [
        { Name = "Tower1"; Description = None; Location = testLocation1 }
        { Name = "Tower2"; Description = None; Location = testLocation2 }
    ]
    let testArtists = [{ Name = "Artist1"; Genre = None }]
    let testMovies = [{ Name = "Movie1"; BoxOffice = None }]

    let dataAccess = testDataAccess testAttractions testArtists testMovies
    let result = travelGuides dataAccess "Tower" 5 |> Async.RunSynchronously

    Assert.Equal(2, List.length result)

// ============================================
// キャッシュのテスト
// ============================================

[<Fact>]
let ``cachedDataAccess はデータをキャッシュする`` () =
    let mutable callCount = 0
    let testLocation = {
        Id = LocationId "Q123"
        Name = "Test City"
        Population = 100000
    }
    let testAttractions = [
        { Name = "Test Tower"; Description = None; Location = testLocation }
    ]

    let countingDataAccess =
        { new IDataAccess with
            member _.FindAttractions(name, ordering, limit) =
                async {
                    callCount <- callCount + 1
                    return
                        testAttractions
                        |> List.filter (fun a -> a.Name.Contains(name))
                        |> List.truncate limit
                }
            member _.FindArtistsFromLocation(_, limit) =
                async { return [] }
            member _.FindMoviesAboutLocation(_, limit) =
                async { return [] }
        }

    let cached = cachedDataAccess countingDataAccess |> Async.RunSynchronously

    // 最初の呼び出し
    let _ = cached.FindAttractions("Test", ByName, 10) |> Async.RunSynchronously
    Assert.Equal(1, callCount)

    // 2回目の呼び出し（キャッシュから）
    let _ = cached.FindAttractions("Test", ByName, 10) |> Async.RunSynchronously
    Assert.Equal(1, callCount) // 呼び出し回数は増えない

    // 異なるパラメータでの呼び出し
    let _ = cached.FindAttractions("Other", ByName, 10) |> Async.RunSynchronously
    Assert.Equal(2, callCount)

// ============================================
// Resource のテスト
// ============================================

[<Fact>]
let ``Resource.make は正しくリソースを作成する`` () =
    let mutable acquired = false
    let mutable released = false

    let resource = Resource.make
                    (fun () -> async { acquired <- true; return "resource" })
                    (fun _ -> async { released <- true })

    let result =
        resource
        |> Resource.useAsync (fun r -> async { return r + "-used" })
        |> Async.RunSynchronously

    Assert.Equal("resource-used", result)
    Assert.True(acquired)
    Assert.True(released)

// ============================================
// サンプルデータのテスト
// ============================================

[<Fact>]
let ``sampleLocation は正しい値を持つ`` () =
    Assert.Equal("Tokyo", sampleLocation.Name)
    Assert.Equal(14000000, sampleLocation.Population)

[<Fact>]
let ``sampleAttraction は正しい値を持つ`` () =
    Assert.Equal("Tokyo Tower", sampleAttraction.Name)
    Assert.True(sampleAttraction.Description.IsSome)

[<Fact>]
let ``sampleArtist は正しい値を持つ`` () =
    Assert.Equal("Hikaru Utada", sampleArtist.Name)

[<Fact>]
let ``sampleMovie は正しい値を持つ`` () =
    Assert.Equal("Lost in Translation", sampleMovie.Name)

